cpp/src/hdbscan/detail/condense.cuh

```diff
@@ -24,11 +24,215 @@
+    if (!internal_nodes.empty()) {
+      std::vector<value_idx> next_queue;
+      for (value_idx node : internal_nodes) {
+        value_idx left  = h_children[node * 2];
+        value_idx right = h_children[node * 2 + 1];
+        next_queue.push_back(left);
+        next_queue.push_back(right);
+      }
+      process_queue = next_queue;
+    } else {
+      process_queue.clear();
+    }
+  }
+}
+
+/**
+ * Performs a level-by-level BFS traversal of the dendrogram
+ * and applies the following logic at each internal node:
+ *
+ * 1. If both children are >= min_cluster_size:
+ *    - Create new cluster labels for both children
+ *    - Add edges from parent to both children
+ *
+ * 2. If both children are < min_cluster_size:
+ *    - Collapse both subtrees
+ *    - Add edges from parent directly to all leaf nodes in both subtrees
+ *
+ * 3. If only left child is < min_cluster_size:
+ *    - Right child inherits parent's label
+ *    - Collapse left subtree, add edges to all its leaf nodes
+ *
+ * 4. If only right child is < min_cluster_size:
+ *    - Left child inherits parent's label
+ *    - Collapse right subtree, add edges to all its leaf nodes
+ *
+ * we ignore subnodes of the collapsed subtree by using a boolean array
+ */
+template <typename value_idx, typename value_t>
+void _build_condensed_hierarchy(const raft::handle_t& handle,
+                                const value_idx* children,
+                                const value_t* delta,
+                                const value_idx* sizes,
+                                int n_leaves,
+                                int min_cluster_size,
+                                rmm::device_uvector<value_idx>& out_parent,
+                                rmm::device_uvector<value_idx>& out_child,
+                                rmm::device_uvector<value_t>& out_lambda,
+                                rmm::device_uvector<value_idx>& out_size)
+{
+  cudaStream_t stream  = handle.get_stream();
+  value_idx root       = 2 * (n_leaves - 1);
+  value_idx n_samples  = n_leaves;
+  value_idx next_label = n_samples + 1;
+
+  // Copy data to host
+  std::vector<value_idx> h_children(2 * (n_leaves - 1));
+  std::vector<value_t> h_delta(n_leaves - 1);
+  std::vector<value_idx> h_sizes(n_leaves - 1);
+
+  raft::copy(h_children.data(), children, 2 * (n_leaves - 1), stream);
+  raft::copy(h_delta.data(), delta, n_leaves - 1, stream);
+  raft::copy(h_sizes.data(), sizes, n_leaves - 1, stream);
+  handle.sync_stream(stream);
+
+  // host side output arrays
+  std::vector<value_idx> h_out_parent;
+  std::vector<value_idx> h_out_child;
+  std::vector<value_t> h_out_lambda;
+  std::vector<value_idx> h_out_sizes;
+
+  // Get BFS ordering from root
+  std::vector<value_idx> node_list;
+  bfs_from_node(root, n_samples, h_children.data(), node_list);
+
+  std::vector<value_idx> relabel(root + 1, 0);
+  std::vector<bool> ignore(node_list.size(), false);
```

Comment:
`std::vector<bool>` is weird, it's like a bitset. That's unnecessary because we do not need the space savings or the overhead of bitwise manipulation. 
```suggestion
  std::vector<uint8_t> ignore(node_list.size(), false);
```